src/ucp/wireup/select.c

```diff
@@ -453,6 +461,65 @@ ucp_wireup_select_transport(ucp_ep_h ep, const ucp_address_entry_t *address_list
     return UCS_OK;
 }
 
+static inline int
+ucp_wireup_can_use_am(const ucp_wireup_select_ctx_t *select_ctx,
+                      uint32_t usage)
+{
+    return (select_ctx->allow_am & usage);
+}
+
+/* Return:
+ * - UCS_OK if suitable transport was found
+ * - UCS_INPROGRESS if suitable transport wasn't found, but
+ *   it can be replaced by emulation over AM
+ * - other error code if an error occured during selection */
+static ucs_status_t
+ucp_wireup_select_lane(ucp_wireup_select_ctx_t *select_ctx,
+                       const ucp_wireup_criteria_t *criteria,
+                       uint64_t tl_bitmap, uint64_t remote_md_map,
+                       uint64_t local_dev_bitmap, uint64_t remote_dev_bitmap,
+                       uint32_t usage, int show_error,
+                       ucp_wireup_select_info_t *select_info)
+{
+    ucp_ep_h ep                 = select_ctx->ep;
+    uint64_t scalable_tl_bitmap = (select_ctx->scalable_tl_bitmap &
+                                   tl_bitmap);
+    ucs_status_t status;
+
+    /* Select lane among scalable transports */
+    if (scalable_tl_bitmap) {
+        status = ucp_wireup_select_transport(ep, select_ctx->address_list,
+                                             select_ctx->address_count, criteria,
+                                             scalable_tl_bitmap, remote_md_map,
+                                             local_dev_bitmap, remote_dev_bitmap,
+                                             0 /* don't log errors */, select_info);
+        if (status == UCS_OK) {
+            ucs_assert(ucp_is_scalable_transport(ep->worker->context,
+                                                 select_info->max_num_eps));
+            return status;
+        }
+
+        /* If we get there, it means that the selected AM transport
```

Comment:
just noticed that the comment is not correct - at this point you do not know that am lane/transport is scalable enough

Reply:
In theory scalable bitmap can contain some transport which is not suitable for AM.
Also you check that AM is really scalable below by this check
`ucp_is_scalable_transport(ep->worker->context, select_ctx->am_info.max_num_eps)`

So the comment is a bit misleading